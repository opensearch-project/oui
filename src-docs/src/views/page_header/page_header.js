/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import {
  OuiPageHeader,
  OuiCode,
  OuiText,
  OuiButton,
} from '../../../../src/components';

export default () => (
  <OuiPageHeader
    pageTitle="Page title"
    tabs={[
      {
        label: 'Tab 1',
        isSelected: true,
      },
      {
        label: 'Tab 2',
      },
    ]}
    description="This description should be describing the current page as depicted by the page title. It has the grow prop set to false on the OuiText block so that it is the proper line length. And it will also never extend beneath the right side content (buttons)."
    rightSideItems={[
      <OuiButton fill>Add something</OuiButton>,
      <OuiButton>Do something</OuiButton>,
    ]}>
    <OuiText>
      <p>
        This custom content (children), on the other hand, exists below the
        content above including below the right side content and therefore will
        stretch beneath them. Unless you set the <OuiCode>alignItems</OuiCode>{' '}
        prop to something other than <OuiCode>top</OuiCode>.
      </p>
    </OuiText>
  </OuiPageHeader>
);
