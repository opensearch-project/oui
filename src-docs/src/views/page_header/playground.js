/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/* eslint-disable @typescript-eslint/no-unused-vars */

import React from 'react';
import { PropTypes } from 'react-view';
import {
  OuiPageTemplate,
  OuiPageHeader,
  OuiButton,
  OuiTabs,
  OuiImage,
} from '../../../../src/components/';
import {
  propUtilityForPlayground,
  simulateFunction,
  generateCustomProps,
  createOptionalEnum,
} from '../../services/playground';

const tabs = `[
  {
    label: 'Tab 1',
    isSelected: true,
  },
  {
    label: 'Tab 2',
  },
]`;

const rightSideItems = `[
  <OuiButton fill>Button 1</OuiButton>,
  <OuiButton>Button 2</OuiButton>,
]`;

// TODO: Try later to build a toggle that allows switching between different types of content to pass
// const rightSideItems =
//   '[<OuiImage url="https://source.unsplash.com/400x200/?Water" height="200" />]';

export const pageHeaderConfig = () => {
  const docgenInfo = Array.isArray(OuiPageHeader.__docgenInfo)
    ? OuiPageHeader.__docgenInfo[0]
    : OuiPageHeader.__docgenInfo;
  const propsToUse = propUtilityForPlayground(docgenInfo.props);

  propsToUse.pageTitle = {
    ...propsToUse.pageTitle,
    type: PropTypes.String,
    value: 'Page title',
  };

  propsToUse.description = {
    ...propsToUse.description,
    value: 'Example of a description.',
    type: PropTypes.String,
  };

  propsToUse.rightSideItems = simulateFunction({
    ...propsToUse.rightSideItems,
    custom: {
      value: rightSideItems,
    },
  });

  propsToUse.tabs = simulateFunction({
    ...propsToUse.tabs,
    custom: {
      value: tabs,
    },
  });

  propsToUse.children = {
    ...propsToUse.children,
    type: PropTypes.ReactNode,
    hidden: false,
  };

  propsToUse.alignItems = createOptionalEnum(propsToUse.alignItems);

  return {
    config: {
      componentName: 'OuiPageHeader',
      props: propsToUse,
      scope: {
        OuiPageHeader,
        OuiButton,
        OuiTabs,
        OuiImage,
      },
      imports: {
        '@opensearch-project/oui': {
          named: ['OuiPageHeader', 'OuiButton', 'OuiTabs', 'OuiImage'],
        },
      },
      customProps: generateCustomProps(['rightSideItems', 'tabs']),
    },
  };
};
