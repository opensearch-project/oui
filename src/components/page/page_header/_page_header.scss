/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/**
 * Note: Bottom margin is added in _page.scss when OuiPage has `paddingSize`
 *       Or it's added in _page_body.scss when OuiPageBody has `paddingSize`
 */

.ouiPageHeader {
  @include ouiPageRestrictWidth;

  width: 100%;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  flex-shrink: 0; // Ensures Safari doesn't shrink beyond contents
}

.ouiPageHeader--bottomBorder {
  border-bottom: $ouiBorderThin;

  &:not(.ouiPageHeader--tabsAtBottom) {
    // Default padding to separate contents (unless there's tabs)
    padding-bottom: map-get($ouiPanelPaddingModifiers, 'paddingLarge');
  }
}

// Uses the same values as OuiPanel
@each $modifier, $amount in $ouiPanelPaddingModifiers {
  .ouiPageHeader--#{$modifier} {
    padding: $amount;

    // Use margin if there are tabs to keep border close to tabs
    &.ouiPageHeader--tabsAtBottom {
      padding-bottom: 0;
    }

    &.ouiPageHeader--tabsAtBottom.ouiPageHeader--bottomBorder {
      margin-bottom: $amount;
    }
  }
}

.ouiPageHeader--top {
  align-items: flex-start;
}

.ouiPageHeader--bottom {
  align-items: flex-end;
}

.ouiPageHeader--stretch {
  align-items: stretch;
}

@include ouiBreakpoint('xs', 's') {
  .ouiPageHeader--responsive {
    flex-direction: column;
  }

  .ouiPageHeader--responsiveReverse {
    flex-direction: column-reverse;
  }
}
