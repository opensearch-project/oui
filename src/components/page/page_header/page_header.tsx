/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { FunctionComponent, HTMLAttributes } from 'react';
import classNames from 'classnames';
import { CommonProps, keysOf } from '../../common';
import {
  OuiPageHeaderContent,
  OuiPageHeaderContentProps,
} from './page_header_content';
import {
  _OuiPageRestrictWidth,
  setPropsForRestrictedPageWidth,
} from '../_restrict_width';

const paddingSizeToClassNameMap = {
  none: null,
  s: 'ouiPageHeader--paddingSmall',
  m: 'ouiPageHeader--paddingMedium',
  l: 'ouiPageHeader--paddingLarge',
};

export const PADDING_SIZES = keysOf(paddingSizeToClassNameMap);

export type OuiPageHeaderProps = CommonProps &
  HTMLAttributes<HTMLElement> &
  OuiPageHeaderContentProps &
  _OuiPageRestrictWidth & {
    /**
     * Adjust the padding.
     * When using this setting it's best to be consistent throughout all similar usages
     */
    paddingSize?: typeof PADDING_SIZES[number];
    /**
     * Adds a bottom border to separate it from the content after
     */
    bottomBorder?: boolean;
  };

export const OuiPageHeader: FunctionComponent<OuiPageHeaderProps> = ({
  className,
  restrictWidth = false,
  paddingSize = 'none',
  bottomBorder,
  style,

  // Page header content shared props:
  alignItems,
  responsive = true,
  children,

  // Page header content only props:
  pageTitle,
  tabs,
  tabsProps,
  description,
  rightSideItems,
  rightSideGroupProps,
  ...rest
}) => {
  const { widthClassName, newStyle } = setPropsForRestrictedPageWidth(
    restrictWidth,
    style
  );

  const classes = classNames(
    'ouiPageHeader',
    paddingSizeToClassNameMap[paddingSize],
    {
      'ouiPageHeader--bottomBorder': bottomBorder,
      'ouiPageHeader--responsive': responsive === true,
      'ouiPageHeader--responsiveReverse': responsive === 'reverse',
      'ouiPageHeader--tabsAtBottom': pageTitle && tabs,
      [`ouiPage--${widthClassName}`]: widthClassName,
    },
    `ouiPageHeader--${alignItems ?? 'center'}`,
    className
  );

  if (!pageTitle && !tabs && !description && !rightSideItems) {
    return (
      <header className={classes} style={newStyle || style} {...rest}>
        {children}
      </header>
    );
  }

  return (
    <header className={classes} style={newStyle || style} {...rest}>
      <OuiPageHeaderContent
        alignItems={alignItems}
        responsive={responsive}
        pageTitle={pageTitle}
        tabs={tabs}
        tabsProps={tabsProps}
        description={description}
        rightSideItems={rightSideItems}
        rightSideGroupProps={rightSideGroupProps}>
        {children}
      </OuiPageHeaderContent>
    </header>
  );
};
