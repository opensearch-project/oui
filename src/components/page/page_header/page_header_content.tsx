/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { FunctionComponent, ReactNode, HTMLAttributes } from 'react';
import classNames from 'classnames';
import { CommonProps } from '../../common';
import { OuiTab, OuiTabs, OuiTabsProps } from '../../tabs';
import { Props as OuiTabProps } from '../../tabs/tab';
import { OuiFlexGroup, OuiFlexItem, OuiFlexGroupProps } from '../../flex';
import { OuiSpacer } from '../../spacer';
import { OuiTitle } from '../../title';
import { OuiText } from '../../text';
import { useIsWithinBreakpoints } from '../../../services/hooks';

export const ALIGN_ITEMS = ['top', 'bottom', 'center', 'stretch'] as const;

// Gets all the tab props including the button or link props
type Tab = OuiTabProps & {
  /**
   * Visible text of the tab
   */
  label: ReactNode;
};

export type OuiPageHeaderContentTitle = {
  /**
   * Wrapped in an `H1` so choose appropriately.
   * A simple string is best
   */
  pageTitle?: ReactNode;
};

export type OuiPageHeaderContentTabs = {
  /**
   * In-app navigation presented as large borderless tabs.
   * Accepts an array of `OuiTab` objects;
   * HELP: This is evaluating to `any[]` in the props table
   */
  tabs?: Tab[];
  /**
   * Any extras to apply to the outer tabs container.
   * Extends `OuiTabs`
   */
  tabsProps?: Omit<OuiTabsProps, 'size' | 'expand' | 'display'>;
};

/**
 * The left side can either be a title with optional description;
 * Or a list of tabs,
 * Or a custom node
 */
type OuiPageHeaderContentLeft = OuiPageHeaderContentTitle &
  OuiPageHeaderContentTabs & {
    /**
     * Position is dependent on existing with a `pageTitle` or `tabs`
     * Automatically get wrapped in a single paragraph tag inside an OuiText block
     */
    description?: string | ReactNode;
  };

export type OuiPageHeaderContentProps = CommonProps &
  HTMLAttributes<HTMLDivElement> &
  OuiPageHeaderContentLeft & {
    /**
     * Set to false if you don't want the children to stack at small screen sizes.
     * Set to `reverse` to display the right side content first for the sack of hierarchy (like global time)
     */
    responsive?: boolean | 'reverse';
    /**
     * Vertical alignment of the left and right side content;
     * Default is `middle` for custom content, but `top` for when `pageTitle` or `tabs` are included
     */
    alignItems?: typeof ALIGN_ITEMS[number];
    /**
     * Pass custom an array of content to this side usually up to 3 buttons.
     * The first button should be primary, usually with `fill` and will be visually displayed as the last item,
     * but first in the tab order
     */
    rightSideItems?: ReactNode[];
    /**
     * Additional OuiFlexGroup props to pass to the container of the `rightSideItems`
     */
    rightSideGroupProps?: Partial<OuiFlexGroupProps>;
    /**
     * Custom children will be rendered before the `tabs` unless no `pageTitle` is present, then it will be the last item
     */
    children?: ReactNode;
  };

export const OuiPageHeaderContent: FunctionComponent<OuiPageHeaderContentProps> = ({
  className,
  pageTitle,
  tabs,
  tabsProps,
  description,
  alignItems = 'top',
  responsive = true,
  rightSideItems,
  rightSideGroupProps,
  children,
  ...rest
}) => {
  const isResponsiveBreakpoint = useIsWithinBreakpoints(
    ['xs', 's'],
    !!responsive
  );

  const classes = classNames('ouiPageHeaderContent');

  let descriptionNode;
  if (description) {
    descriptionNode = (
      <>
        {(pageTitle || tabs) && <OuiSpacer />}
        <OuiText grow={false}>
          <p>{description}</p>
        </OuiText>
      </>
    );
  }

  let pageTitleNode;
  if (pageTitle) {
    pageTitleNode = (
      <OuiTitle size="l">
        <h1>{pageTitle}</h1>
      </OuiTitle>
    );
  }

  let tabsNode;
  if (tabs) {
    const renderTabs = () => {
      return tabs.map((tab, index) => {
        const { label, ...tabRest } = tab;
        return (
          <OuiTab key={index} {...tabRest}>
            {label}
          </OuiTab>
        );
      });
    };

    tabsNode = (
      <>
        {pageTitleNode && <OuiSpacer />}
        <OuiTabs {...tabsProps} display="condensed" size="l">
          {renderTabs()}
        </OuiTabs>
      </>
    );
  }

  const childrenNode = children && (
    <>
      <OuiSpacer />
      {children}
    </>
  );

  let bottomContentNode;
  if (childrenNode || (tabsNode && pageTitleNode)) {
    bottomContentNode = (
      <div className="ouiPageHeaderContent__bottom">
        {childrenNode}
        {pageTitleNode && tabsNode}
      </div>
    );
  }

  /**
   * The left side order depends on if a `pageTitle` was supplied.
   * If not, but there are `tabs`, then the tabs become the page title
   */
  let leftSideOrder;
  if (tabsNode && !pageTitleNode) {
    leftSideOrder = (
      <>
        {tabsNode}
        {descriptionNode}
      </>
    );
  } else {
    leftSideOrder = (
      <>
        {pageTitleNode}
        {descriptionNode}
      </>
    );
  }

  let rightSideFlexItem;
  if (rightSideItems && rightSideItems.length) {
    const wrapWithFlex = () => {
      return rightSideItems.map((item, index) => {
        return (
          <OuiFlexItem grow={false} key={index}>
            {item}
          </OuiFlexItem>
        );
      });
    };

    rightSideFlexItem = (
      <OuiFlexItem grow={false}>
        <OuiFlexGroup
          wrap
          responsive={false}
          {...rightSideGroupProps}
          className={classNames(
            'ouiPageHeaderContent__rightSideItems',
            rightSideGroupProps?.className
          )}>
          {wrapWithFlex()}
        </OuiFlexGroup>
      </OuiFlexItem>
    );
  }

  return alignItems === 'top' || isResponsiveBreakpoint ? (
    <div className={classes} {...rest}>
      <OuiFlexGroup
        responsive={!!responsive}
        className="ouiPageHeaderContent__top"
        alignItems="flexStart"
        gutterSize="l">
        {isResponsiveBreakpoint && responsive === 'reverse' ? (
          <>
            {rightSideFlexItem}
            <OuiFlexItem>{leftSideOrder}</OuiFlexItem>
          </>
        ) : (
          <>
            <OuiFlexItem>{leftSideOrder}</OuiFlexItem>
            {rightSideFlexItem}
          </>
        )}
      </OuiFlexGroup>
      {bottomContentNode}
    </div>
  ) : (
    <div className={classes} {...rest}>
      <OuiFlexGroup
        responsive={!!responsive}
        className="ouiPageHeaderContent__top"
        alignItems={alignItems === 'bottom' ? 'flexEnd' : alignItems}
        gutterSize="l">
        <OuiFlexItem>
          {leftSideOrder}
          {bottomContentNode}
        </OuiFlexItem>
        {rightSideFlexItem}
      </OuiFlexGroup>
    </div>
  );
};
