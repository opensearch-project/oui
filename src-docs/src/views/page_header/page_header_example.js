/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Link } from 'react-router-dom';

import { GuideSectionTypes } from '../../components';

import {
  OuiCode,
  OuiPageHeader,
  OuiPageHeaderSection,
} from '../../../../src/components';

import { pageHeaderConfig } from './playground';

import PageHeader from './page_header';
const pageHeaderSource = require('!!raw-loader!./page_header');

import PageHeaderTabs from './page_header_tabs';
const pageHeaderTabsSource = require('!!raw-loader!./page_header_tabs');

import PageHeaderCustom from './page_header_custom';
import { OuiText } from '../../../../src/components/text';
const pageHeaderCustomSource = require('!!raw-loader!./page_header_custom');

export const PageHeaderExample = {
  title: 'Page header',
  intro: (
    <OuiText>
      <p>
        While the <strong>OuiPageHeader</strong> component can be placed
        anywhere within your page layout, we recommend using it within the{' '}
        <Link to="/layout/page">
          <strong>OuiPageTemplate</strong>
        </Link>{' '}
        component by passing the configuration props as its{' '}
        <OuiCode>pageHeader</OuiCode>.
      </p>
    </OuiText>
  ),
  sections: [
    {
      source: [
        {
          type: GuideSectionTypes.JS,
          code: pageHeaderSource,
        },
      ],
      playground: pageHeaderConfig,
      text: (
        <>
          <p>
            <strong>OuiPageHeader</strong> provides props for opinionated,
            consistent formatting of your header. Any combination of{' '}
            <OuiCode>pageTitle</OuiCode>, <OuiCode>description</OuiCode>,{' '}
            <OuiCode>tabs</OuiCode>, or any <OuiCode>children</OuiCode> will
            adjust the layout as needed.
          </p>
          <p>
            An additional prop <OuiCode>rightSideItems</OuiCode> allows for a
            simple <strong>array of nodes</strong> which will layout in a
            flexbox row. This is commonly used for adding multiple buttons, of
            which, at least one should be primary (or{' '}
            <OuiCode language="ts">{'fill="true"'}</OuiCode>). These items are
            also displayed in <strong>reverse order</strong> so that the first
            and primary array item will be displayed on the far right.
          </p>
          <p>
            You can further adjust the display of these content types by
            utilizing <OuiCode>alignItems</OuiCode> for adjusting the vertical
            alignment of the two sides, and <OuiCode>responsiveOrder</OuiCode>{' '}
            to determine which content side to display first on smaller screens.
          </p>
        </>
      ),
      demo: <PageHeader />,
      props: { OuiPageHeader },
      snippet: `<OuiPageHeader
  pageTitle="Page title"
  tabs={[
    { label:"Tab 1", isSelected: true },
    { label:"Tab 2" }
  ]}
  description="Example of a description."
  rightSideItems={[
    <OuiButton fill>Button 1</OuiButton>,
    <OuiButton>Button 2</OuiButton>
  ]}
/>`,
    },
    {
      title: 'Tabs in the page header',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: pageHeaderTabsSource,
        },
      ],
      text: (
        <>
          <p>
            When supplying <OuiCode>tabs</OuiCode> without a{' '}
            <OuiCode>pageTitle</OuiCode>, <strong>OuiPageHeader</strong> will
            promote those tabs as if they are the page title. This means that
            any <OuiCode>description</OuiCode> or <OuiCode>children</OuiCode>{' '}
            will sit <strong>below</strong> the tabs.
          </p>
        </>
      ),
      demo: <PageHeaderTabs />,
      props: { OuiPageHeader },
      snippet: `<OuiPageHeader
  tabs={[
    { label:"Tab 1", isSelected: true },
    { label:"Tab 2" }
  ]}
  description="Example of a description."
/>`,
    },
    {
      title: 'Customizing the page header',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: pageHeaderCustomSource,
        },
      ],
      text: (
        <>
          <p>
            The page header content props are helpful props to push content into
            established Elastic page layout patterns. They are completely
            optional and by design, inflexible. If you need a layout that does
            not match these patterns you can pass in your own{' '}
            <OuiCode>children</OuiCode> utilizing the{' '}
            <strong>OuiPageHeaderSection</strong> components.
          </p>
        </>
      ),
      demo: <PageHeaderCustom />,
      props: { OuiPageHeader, OuiPageHeaderSection },
    },
  ],
};
