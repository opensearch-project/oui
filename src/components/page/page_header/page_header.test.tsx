/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test';

import { OuiPageHeader, OuiPageHeaderProps } from './page_header';
import { ALIGN_ITEMS } from './page_header_content';

export const tabs: OuiPageHeaderProps['tabs'] = [
  {
    label: 'Tab 1',
    isSelected: true,
  },
  {
    label: 'Tab 2',
  },
];

export const rightSideItems: OuiPageHeaderProps['rightSideItems'] = [
  <button>Button 1</button>,
  <button>Button 2</button>,
];

describe('OuiPageHeader', () => {
  test('is rendered', () => {
    const component = render(
      <OuiPageHeader {...requiredProps}>Anything</OuiPageHeader>
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    describe('page content props are passed down', () => {
      test('is rendered', () => {
        const component = render(
          <OuiPageHeader
            pageTitle="Page title"
            tabs={tabs}
            tabsProps={requiredProps}
            description="Description"
            rightSideItems={rightSideItems}
            rightSideGroupProps={{ responsive: true, ...requiredProps }}>
            <p>Anything</p>
          </OuiPageHeader>
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('alignItems', () => {
      ALIGN_ITEMS.forEach((alignment) => {
        it(`${alignment} is rendered`, () => {
          const component = render(
            <OuiPageHeader
              pageTitle="Page title"
              rightSideItems={rightSideItems}
              alignItems={alignment}
            />
          );

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('responsive', () => {
      test('is rendered as false', () => {
        const component = render(<OuiPageHeader responsive={false} />);

        expect(component).toMatchSnapshot();
      });

      test('is rendered as reverse', () => {
        const component = render(<OuiPageHeader responsive={'reverse'} />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('restrictWidth', () => {
      test('is rendered as true', () => {
        const component = render(<OuiPageHeader restrictWidth={true} />);

        expect(component).toMatchSnapshot();
      });

      test('is rendered as custom', () => {
        const component = render(<OuiPageHeader restrictWidth={100} />);

        expect(component).toMatchSnapshot();
      });
    });
  });
});
