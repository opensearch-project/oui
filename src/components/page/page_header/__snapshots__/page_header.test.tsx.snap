// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiPageHeader is rendered 1`] = `
<header
  aria-label="aria-label"
  class="ouiPageHeader ouiPageHeader--responsive ouiPageHeader--center testClass1 testClass2"
  data-test-subj="test subject string"
>
  Anything
</header>
`;

exports[`OuiPageHeader props alignItems bottom is rendered 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPageHeader--bottom"
>
  <div
    class="ouiPageHeaderContent"
  >
    <div
      class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--alignItemsFlexEnd ouiFlexGroup--directionRow ouiFlexGroup--responsive ouiPageHeaderContent__top"
    >
      <div
        class="ouiFlexItem"
      >
        <h1
          class="ouiTitle ouiTitle--large"
        >
          Page title
        </h1>
      </div>
      <div
        class="ouiFlexItem ouiFlexItem--flexGrowZero"
      >
        <div
          class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--directionRow ouiFlexGroup--wrap ouiPageHeaderContent__rightSideItems"
        >
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 1
            </button>
          </div>
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 2
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`OuiPageHeader props alignItems center is rendered 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPageHeader--center"
>
  <div
    class="ouiPageHeaderContent"
  >
    <div
      class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--alignItemsCenter ouiFlexGroup--directionRow ouiFlexGroup--responsive ouiPageHeaderContent__top"
    >
      <div
        class="ouiFlexItem"
      >
        <h1
          class="ouiTitle ouiTitle--large"
        >
          Page title
        </h1>
      </div>
      <div
        class="ouiFlexItem ouiFlexItem--flexGrowZero"
      >
        <div
          class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--directionRow ouiFlexGroup--wrap ouiPageHeaderContent__rightSideItems"
        >
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 1
            </button>
          </div>
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 2
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`OuiPageHeader props alignItems stretch is rendered 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPageHeader--stretch"
>
  <div
    class="ouiPageHeaderContent"
  >
    <div
      class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--directionRow ouiFlexGroup--responsive ouiPageHeaderContent__top"
    >
      <div
        class="ouiFlexItem"
      >
        <h1
          class="ouiTitle ouiTitle--large"
        >
          Page title
        </h1>
      </div>
      <div
        class="ouiFlexItem ouiFlexItem--flexGrowZero"
      >
        <div
          class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--directionRow ouiFlexGroup--wrap ouiPageHeaderContent__rightSideItems"
        >
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 1
            </button>
          </div>
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 2
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`OuiPageHeader props alignItems top is rendered 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPageHeader--top"
>
  <div
    class="ouiPageHeaderContent"
  >
    <div
      class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--alignItemsFlexStart ouiFlexGroup--directionRow ouiFlexGroup--responsive ouiPageHeaderContent__top"
    >
      <div
        class="ouiFlexItem"
      >
        <h1
          class="ouiTitle ouiTitle--large"
        >
          Page title
        </h1>
      </div>
      <div
        class="ouiFlexItem ouiFlexItem--flexGrowZero"
      >
        <div
          class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--directionRow ouiFlexGroup--wrap ouiPageHeaderContent__rightSideItems"
        >
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 1
            </button>
          </div>
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 2
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;

exports[`OuiPageHeader props page content props are passed down is rendered 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPageHeader--tabsAtBottom ouiPageHeader--center"
>
  <div
    class="ouiPageHeaderContent"
  >
    <div
      class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--alignItemsFlexStart ouiFlexGroup--directionRow ouiFlexGroup--responsive ouiPageHeaderContent__top"
    >
      <div
        class="ouiFlexItem"
      >
        <h1
          class="ouiTitle ouiTitle--large"
        >
          Page title
        </h1>
        <div
          class="ouiSpacer ouiSpacer--l"
        />
        <div
          class="ouiText ouiText--medium ouiText--constrainedWidth"
        >
          <p>
            Description
          </p>
        </div>
      </div>
      <div
        class="ouiFlexItem ouiFlexItem--flexGrowZero"
      >
        <div
          aria-label="aria-label"
          class="ouiFlexGroup ouiFlexGroup--gutterLarge ouiFlexGroup--directionRow ouiFlexGroup--responsive ouiFlexGroup--wrap ouiPageHeaderContent__rightSideItems testClass1 testClass2"
          data-test-subj="test subject string"
        >
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 1
            </button>
          </div>
          <div
            class="ouiFlexItem ouiFlexItem--flexGrowZero"
          >
            <button>
              Button 2
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ouiPageHeaderContent__bottom"
    >
      <div
        class="ouiSpacer ouiSpacer--l"
      />
      <p>
        Anything
      </p>
      <div
        class="ouiSpacer ouiSpacer--l"
      />
      <div
        aria-label="aria-label"
        class="ouiTabs ouiTabs--condensed ouiTabs--large testClass1 testClass2"
        data-test-subj="test subject string"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="ouiTab ouiTab-isSelected"
          role="tab"
          type="button"
        >
          <span
            class="ouiTab__content"
          >
            Tab 1
          </span>
        </button>
        <button
          aria-selected="false"
          class="ouiTab"
          role="tab"
          type="button"
        >
          <span
            class="ouiTab__content"
          >
            Tab 2
          </span>
        </button>
      </div>
    </div>
  </div>
</header>
`;

exports[`OuiPageHeader props responsive is rendered as false 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--center"
/>
`;

exports[`OuiPageHeader props responsive is rendered as reverse 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsiveReverse ouiPageHeader--center"
/>
`;

exports[`OuiPageHeader props restrictWidth is rendered as custom 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPage--restrictWidth-custom ouiPageHeader--center"
  style="max-width:100px"
/>
`;

exports[`OuiPageHeader props restrictWidth is rendered as true 1`] = `
<header
  class="ouiPageHeader ouiPageHeader--responsive ouiPage--restrictWidth-default ouiPageHeader--center"
/>
`;
